field/subfields.go

```diff
@@ -0,0 +1,223 @@
+// in the specification.
+// NOTE: Subfields do not support encoding and padding. Therefore, users should
+// only pass None or nil values for these types. Passing any other value will
+// result in a panic.
+func (f *Subfields) SetSpec(spec *Spec) {
+	if err := validateSubfieldsSpec(spec); err != nil {
+		panic(err)
+	}
+	f.spec = spec
+	f.idToFieldMap = spec.CreateMessageFields()
+	f.orderedFieldIDs = orderedKeys(f.idToFieldMap)
+}
+
+// SetData traverses through fields provided in the data parameter matches them
+// with their spec definition and calls SetData(...) on each spec field with the
+// appropriate data field.
+//
+// A valid input is as follows:
+//
+//		type SubfieldData struct {
+//			F1 *String
+//			F2 *String
+//			F3 *Numeric
+//		}
+//
+func (f *Subfields) SetData(data interface{}) error {
+	f.data = data
+
+	if f.data == nil {
+		return nil
+	}
+
+	str := reflect.ValueOf(data)
+	strKind := str.Kind()
+	if strKind == reflect.Ptr || strKind == reflect.Interface {
+		// get the struct
+		str = str.Elem()
+	}
+
+	if str.Kind() != reflect.Struct {
+		return fmt.Errorf("failed to set data as struct is expected, got: %s", reflect.TypeOf(str).Kind())
+	}
+
+	for i, specField := range f.idToFieldMap {
+		fieldName := fmt.Sprintf("F%d", i)
+
+		// get the struct field
+		dataField := str.FieldByName(fieldName)
+
+		// no data field was found with this name
+		if dataField == (reflect.Value{}) {
+			continue
+		}
+
+		if dataField.IsNil() {
+			dataField.Set(reflect.New(dataField.Type().Elem()))
+		}
+		if err := specField.SetData(dataField.Interface()); err != nil {
+			return fmt.Errorf("failed to set data for field %d: %w", i, err)
+		}
+	}
+	return nil
+}
+
+// Pack deserialises data held by the receiver (via SetData)
+// into bytes and returns an error on failure.
+func (f *Subfields) Pack() ([]byte, error) {
+	packed := []byte{}
+	for _, id := range f.orderedFieldIDs {
+		packedBytes, err := f.idToFieldMap[id].Pack()
+		if err != nil {
+			return nil, fmt.Errorf("failed to pack subfield %d: %v", id, err)
+		}
+		packed = append(packed, packedBytes...)
+	}
+
+	packedLength, err := f.spec.Pref.EncodeLength(f.spec.Length, len(packed))
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode length: %v", err)
+	}
+
+	return append(packedLength, packed...), nil
+}
+
+// Unpack takes in a byte array and serialises them into the receiver's
+// subfields. An offset (unit depends on encoding and prefix values) is
+// returned on success. A non-nil error is returned on failure.
+func (f *Subfields) Unpack(data []byte) (int, error) {
+	dataLen, err := f.spec.Pref.DecodeLength(f.spec.Length, data)
+	if err != nil {
+		return 0, fmt.Errorf("failed to decode length: %v", err)
+	}
+
+	offset := f.spec.Pref.Length()
+	for _, id := range f.orderedFieldIDs {
+		fl := f.idToFieldMap[id]
+		read, err := fl.Unpack(data[offset:])
+		if err != nil {
+			return 0, fmt.Errorf("failed to unpack subfield %d: %v", id, err)
+		}
+		offset += read
+	}
```

Comment:
nit: it bothers me that this is essentially the same as `func (f *Subfields) SetBytes(data []byte) error {`. We could call some common non-exported function in both possibly, where this uses `data[offset:]` when calling the function, and the other just ignores the returned length

Reply:
good point!